sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala

```diff
@@ -1847,6 +1847,36 @@ object GpuOverrides {
         override def convertToGpu(): Expression =
           sortOrder.withNewChildren(childExprs.map(_.convertToGpu()))
       }),
+    expr[PivotFirst](
+      "PivotFirst operator",
+      ExprChecks.groupByAgg(
```

Comment:
So I was able to make PivotFirst work as a reduction (no keys in the HashAggregate). Spark does not make it simple/obvious how to do it but.

```
scala> df.groupBy().pivot("a").max("b").show
+---+---+---+---+
|  0|  1| 12|123|
+---+---+---+---+
|  7|  6|  8|  5|
+---+---+---+---+


scala> df.groupBy().pivot("a").max("b").explain
== Physical Plan ==
HashAggregate(keys=[], functions=[pivotfirst(a#2519, max(`b`)#6016, 0, 1, 12, 123, 0, 0)])
+- Exchange SinglePartition, true, [id=#309]
   +- HashAggregate(keys=[], functions=[partial_pivotfirst(a#2519, max(`b`)#6016, 0, 1, 12, 123, 0, 0)])
      +- *(2) HashAggregate(keys=[a#2519], functions=[max(b#2520)])
         +- Exchange hashpartitioning(a#2519, 200), true, [id=#304]
            +- *(1) HashAggregate(keys=[a#2519], functions=[partial_max(b#2520)])
               +- *(1) LocalTableScan [a#2519, b#2520]
```

I think it might be time to have a bit more complex API, that might end up being a bit cleaner in the long run probably using a builder to be able to set 

```
def aggBuilder(
      outputCheck: TypeSig,
      sparkOutputSig: TypeSig,
      paramCheck: Seq[ParamCheck] = Seq.empty,
      repeatingParamCheck: Option[RepeatingParamCheck] = None): ExprChecksBuilder = {
 ...
}

class ExprChecksBuilder(...) {
    def rapidsAndSparkSupportsGroupBy ...
    def rapidsAndSparkSupportsReduction ...
    def sparkSupportsGroupByAndReduction ...
    def rapidsAndSparkSupportsWindow ...
    def sparkSupportsWindow ...
    def build(): ExprChecks
}
      
```

Otherwise if we keep doing what we have been doing we will have 2^5 different aggregation APIs for type checking.

Reply:
Thanks, @revans2. Updated the PR to support queries with reduction only on Pivot. Added integration tests as well. 
I have created a follow on issue to refactor ExprChecks https://github.com/NVIDIA/spark-rapids/issues/2152.
PTAL.